processes/record_clusterer.py

```diff
@@ -0,0 +1,188 @@
+
+        matched_record_ids = set()
+        checked_ids = set()
+
+        for match_distance in range(0, self.MAX_MATCH_DISTANCE):
+            matched_records = self._get_matched_records(list(ids_to_check), matched_record_ids.copy())
+
+            if not matched_records:
+                break
+
+            checked_ids.update(ids_to_check)
+            ids_to_check.clear()
+
+            for matched_record_title, matched_record_id, matched_record_identifiers in matched_records:
+                if not matched_record_title:
+                    logger.warning('Invalid title found in matched records')
+                    continue
+
+                tokenized_matched_record_title = self._tokenize_title(matched_record_title)
+
+                if match_distance > 0 and not self._titles_overlap(tokenized_record_title, tokenized_matched_record_title):
+                    continue
+
+                ids_to_check.update({id for id in matched_record_identifiers 
+                                     if re.search(self.IDENTIFIERS_TO_MATCH, id) and id not in checked_ids})
+                matched_record_ids.add(matched_record_id)
+
+        if len(matched_record_ids) > self.CLUSTER_SIZE_LIMIT:
+            raise Exception(f'Records matched is greater than {self.CLUSTER_SIZE_LIMIT}')
+
+        return list(matched_record_ids)
+
+    def _get_matched_records(self, identifiers: list[str], already_matched_record_ids: list[str]):
+        batch_size = 100
+        matched_records = []
+
+        for i in range(0, len(identifiers), batch_size):
+            id_batch = self._format_identifiers(identifiers[i:i+batch_size])
+
+            try:
+                records = (
+                    self.db_manager.session.query(
+                        Record.title, Record.id, Record.identifiers)
+                    .filter(~Record.id.in_(already_matched_record_ids))
+                    .filter(Record.identifiers.overlap(id_batch))
+                    .filter(Record.title.isnot(None))
+                    .all()
+                )
+
+                matched_records.extend(records)
+            except DataError:
+                logger.exception('Unable to get matching records')
+
+        return matched_records
+
+    def _create_work_from_editions(self, editions: list, records: list[Record]):
+        record_manager = SFRRecordManager(self.db_manager.session, self.constants['iso639'])
+
+        work_data = record_manager.buildWork(records, editions)
+
+        record_manager.saveWork(work_data)
+
+        stale_work_ids = record_manager.mergeRecords()
+
+        return record_manager.work, stale_work_ids
+
+    def _index_works_in_elastic_search(self, works: Work):
+        work_documents = []
+
+        for work in works:
+            elastic_manager = SFRElasticRecordManager(work)
+            elastic_manager.getCreateWork()
+            work_documents.append(elastic_manager.work)
+
+        self.elastic_search_manager.saveWorkRecords(work_documents)
```

Comment:
We can get rid of the for loop!

We may want to add a TODO here to save a single work!